### ISSUE TITLE
Classes and IDs

### ISSUE THREAD

**Issue by saivan** [asciimath/asciimathml#68]
I was wondering if it would be possible to include classes or ids in the asciimath spec. This would make it easy to highlight key parts of an equation or to select particular symbols. You could utilise something like the limit pattern to achieve this, perhaps something like:

```
    id_(choose-me)(a^2 + b^2)
    class_(highlight-me)(a^2 + b^2)
```

If we had this, adding color to parts of equations or styling parts of equations would be pretty easy. By default, the id or the class would just be applied to the next character, but the brackets would allow you to apply it to the whole statement.

If this is implemented, we could then also take this a step further and implement color very easily; because it would follow this same pattern.

I'd be interested in taking a crack at implementing this if you think it would be useful :)

**Comment by drlippman** [comment]
While this might be possible in AsciiMathML.js, I'm not sure how well it would translate to MathJax's rendering approaches. 

FWIW, there's already a color command that can be used to set the color of parts of an expression, like `x^2+color(red)(y^2)=4`.

**Comment by saivan** [comment]
Well MathJax already has a class directive in latex, its just `\class{choose-me}{a^2 + b^2}`. I'm not so sure about id's to be totally honest, but classes certainly seem to work.

If you've already got color, then shouldn't the id and class just follow the same patterns?

**Comment by drlippman** [comment]
Ahh, interesting.  In that case, it probably wouldn't be too hard to add a new binary type symbol, like the `color` symbol (but with `mrow` as the tag, which is how MathJax appears to handle the class in the MML output).  Follow that up with a code branch in `AMparseSexpr` after the color branch. You could probably copy-paste the color code block, and change the `setAttribute` to set the `class`.

**Comment by pkra** [comment]
>  I'm not sure how well it would translate to MathJax's rendering approaches.

While both class and id (and most other attributes) would be passed along, the CSS associated with them might not have the desired (or any) effect. 

Simple things like changing colors usually work though.

**Comment by saivan** [comment]
So is this something we should try?
Would you like me to try to implement it and make a pr?

**Comment by pkra** [comment]
Just another side comment: if this proposal is accepted by the asciimath maintainers, I would recommend to look at the syntax MathJax's (non-standard) TeX extension `html`, http://docs.mathjax.org/en/latest/tex.html#html in mind.

The reason to recommend this is that we (at MathJax) are considering switching to an ascii-to-tex-to-mathjax conversion in MathJax v3 (replacing the current ascii-to-mathml-to-mathjax conversion).

**Comment by jipsen** [comment]
Interesting discussion. I'm not sure if I am still an official asciimath maintainer (since David Lippman has been doing it for many years now), but I support the suggestion to include classes and ids in asciimath. Also changing the conversion to ascii-to-tex-to-mathjax seems very reasonable to me (not sure how much work it would require).

**Comment by drlippman** [comment]
An AsciiMath-to-TeX converter already exists (it's in the repo), so that wouldn't take much work.

BTW, Peter, there's a couple PRs I'd love your input on before merging in, if you have the time.

**Comment by saivan** [comment]
So is MathJax going to deprecate Mathml then? I think I might just go ahead and implement this when I get the time, but as you all said, this *hopefully* won't be too difficult. My only hazy point is whether this would also need changes to MathJax then. I think something like this could make mathml much more attractive, as you'd have a really easy way to highlight key points in an equation :)

**Comment by pkra** [comment]
> So is MathJax going to deprecate Mathml then? 

No.

**Comment by drlippman** [comment]
@saivan, I just realized we never directly answered one of your questions.

> Would you like me to try to implement it and make a pr?

Sure, go for it :)  I'd imagine the implementation will be similar to how "color" was implemented.

**Comment by saivan** [comment]
Awesome awesome, I have a bit of a busy few weeks ahead, but I'll get to it as soon as I can :) Thanks @drlippman

**Comment by saivan** [comment]
So I finally have some time to try this, and I've hit a bit of a roadblock looking at all of the different files here :P (maybe we can document the purpose of everything in the README). 

I reasoned that the only file that would directly need this is the `ASCIIMathTeXImg.js` file, since it would just convert the asciimath to latex directly (which would support what @pkra said about mathjax v3)  - which we could then use in MathJax. But then theres also `jax.js`, which doesn't seem to have any tests in the repo (unless I'm missing them)

So I'm going to go ahead and change `ASCIIMathTeXImg.js` first, but it looks like there is a little bit of code duplication throughout this repo (which might make a good case for potentially using a bundler?)

**Comment by saivan** [comment]
Actually, I think I need to do `ASCIIMathML.js` primarily so that it can work with current versions of mathjax. But yes, any input that can shed a little more light on this would be most appreciated!

**Comment by saivan** [comment]
Okay so I think I got this working here
https://github.com/asciimath/asciimathml/pull/86

Just check it all out for me, but it lets me do some pretty cool (and dare I say useful) stuff :)

**Comment by drlippman** [comment]
Closed thanks to saivan's PR in #86.
